### DISCUSSION TITLE
Mellinger paper and Thrust

### DISCUSSION THREAD

**Discussion by magalhaesw** [bitcraze/discussions#488] [Q&A] [answered]
Dear Group,
Id like to ask for help. I am working on writing my customized onboard controller. I am studying the script "controller_mellinger.c" and the paper "Minimum Snap Trajectory Generation and Control for Quadrotors " . I have a question about equation (2) of this paper shown below:

$$\begin{bmatrix} Ftotal\\ Mx\\ My\\ Mz\end{bmatrix}^\mathrm{T} = \begin{bmatrix} K_F & K_F & K_F&K_F\\
         0 & K_{F}*L & 0 & -K_{F}*L\\
         -K_{F}*L & 0 & K_{F}*L & 0\\
         K_{M} & -K_{M} & K_{M} & -K_{M}\\
        \end{bmatrix}\begin{bmatrix}\omega_{1}^2\\ \omega_{2}^2\\ \omega_{3}^2\\ \omega_{4}^2 \end{bmatrix}^\mathrm{T}$$

where $\omega_{i}$ is the angular speed of each motor "i" controlled via PWM.

This equation relates the controllers to motor speed, i.e., I set the values of $\omega_{i}$ in rpm (PWM) according to the values of the controllers $[Ftotal, Mx, My, Mz]$ .

Please,
1) What portion of the code is the above equation solved / implemented?  I mean, where the values of $\omega_{i}$ (in PWM) are set based on the controllers.
2) Where are the values of  ( $K_{F},K_{M},L$ ) defined in the code ?  

I looked at **"power_distribution_quadrotor.c"** but I did not find the answers for these questions.

Thank you so much for your help,
Jose

**Comment by hmllr** [comment, marked as answer]
Hi :)
From a quick look at the paper it seems as if you should be looking at chapter 4 (control) instead of chapter 2 (model) for the formulas actually used to implement the controller.
Hope this helps at least a little bit,
Hanna

**Reply by magalhaesw** [reply]
Hi Hanna !
Thanks for your help. 
I asked this question to understand further how to control the motors to produce the correct Thrust.

I am working on coding a LQR for hover condition using the full state and the input as described in the Mellinger paper, i.e.,  
**[u_1, u_2, u_3, u_4]  = [Ftotal,  Mx, My,  Mz].** 
The values of input is given by u = k*(x - xref) + uref,  where "k" is given by the linearization of the quadrotor dynamics described in the Paper.

Then, I added:
 _**control->thrust = u1;**  
**control->roll = u2;  
control-> pitch = u3;  
control-> yaw = u4.**_

However, the value of u1 is too low and nothing happens.

Based on the values of thrust for PID controller, I have tried  "control->thrust = (10^5)*u1".
The crazyflies takes off but it "crashes".

Thanks,
     Jose.

**Reply by hmllr** [reply]
Hmm, if it doesn´t take off there seems to be a mistake in the implementation of the dynamics.

In the Mellinger controller in the Crazyflie u1 is computed here: [https://github.com/bitcraze/crazyflie-firmware/blob/master/src/modules/src/controller_mellinger.c#L169](url) and [https://github.com/bitcraze/crazyflie-firmware/blob/master/src/modules/src/controller_mellinger.c#L213](url) following the formula in the beginning of the Control chapter of the paper. u2-4 are computed using formula (8).  So if you want to follow the paper maybe you can just re-use the same code?

**Reply by magalhaesw** [reply]
Before flying, I simulated the dynamics of CF2 in Python.  
In simulation, for any given initial state, the system converges to (x,y,z) = (0,0,z_target). 
For this simulation , I am using the following parameters:   **gravity = 9.81;  mass = 0.0313;  I = np.diag([1.4194e-05, 1.4089e-05,  2.9741e-05]).**

I will take a look at the implementation of the dynamics.
Maybe these parameters are not **quite correct**.

Using the gains (matrix "k") from the simulation, I set the thrust (u1) like here: https://github.com/bitcraze/crazyflie-firmware/blob/master/src/modules/src/controller_mellinger.c#L295
and the body Moments (u2, u3, u4) like here:  https://github.com/bitcraze/crazyflie-firmware/blob/master/src/modules/src/controller_mellinger.c#L305

I will keep studying the Mellinger controller to debug my code for LQR.
I will consider to re-use the same code.

Thanks for your help Hanna !

Please, I have a basic question.  
The CF2 in a hover, I might have F = (Fx, Fy, Fz) = (0, 0, mass*gravity) = (0, 0, 0.0313 * 9.81) = (0, 0, 0.3070).

So, in a hover,  Do I have to set the thrust like :   "control->thrust  = 0.3070 " ?
This value seems to be very low as nothing happens with CF2 for "control->thrust  = 0.3070 ", 

Thanks for the clarification.
Jose

**Reply by hmllr** [reply]
If you set the thrust to 0.307 nothing at all should happen - the values for the PWM are in uint16 at the end. You are forgetting the massThrust parameter that is multiplied with the raw thrust in the Mellinger controller implementation of the Crazyflie.

**Reply by magalhaesw** [reply]
Thanks Hanna !!
Thank you so much for your help.
